src/napari/_vispy/layers/tiled_image.py

```diff
@@ -0,0 +1,104 @@
+        super().__init__()
+
+        self.set_data(data)
+
+    def set_data(self, data: np.ndarray) -> None:
+        tiles = make_tiles(data, self.tile_size)
+        self.offsets = [of for of, _ in tiles]
+        self.data = data
+        # if the correct number of tiles already exist, just update their data
+        # otherwise, delete all existing tiles and create new ones
+        if len(self.adopted_children) == len(tiles):
+            for ch, (_, dat) in zip(self.adopted_children, tiles, strict=True):
+                ch.set_data(dat)
+        else:
+            for child in self.adopted_children:
+                child.parent = None
+            self.adopted_children = [
+                Image(
+                    data=dat,
+                    parent=self,
+                    texture_format=self.texture_format,
+                )
+                for _, dat in tiles
+            ]
+            for ch, offset in zip(
+                self.adopted_children, self.offsets, strict=True
+            ):
+                ch.transform = STTransform(translate=offset + (0,))
+
+    def set_gl_state(self, *args: Any, **kwargs: Any) -> None:
+        for child in self.adopted_children:
+            child.set_gl_state(*args, **kwargs)
+
+    def __getattr__(self, name: str) -> Any:
+        if (
+            name in ['cmap', 'clim', 'opacity', 'gamma', 'events']
+            and len(self.adopted_children) > 0
+        ):
+            return getattr(self.adopted_children[0], name)
+        return self.__getattribute__(name)
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        if name in ['cmap', 'clim', 'opacity', 'gamma']:
+            for child in self.adopted_children:
+                setattr(child, name, value)
+        else:
+            super().__setattr__(name, value)
+
+
+def make_tiles(
+    image: np.ndarray, tile_size: int
+) -> list[tuple[tuple[int, int], np.ndarray]]:
+    """
+    Splits a large image (grayscale or RGB) into tiles and creates Image visuals for each tile.
+    Supports input shapes (H, W) for grayscale or (H, W, 3) for RGB.
+    """
+    if image.ndim == 2:
+        h, w = image.shape
+    elif image.ndim == 3 and image.shape[2] == 3:
+        h, w, _ = image.shape
+
+    # Calculate number of tiles needed
+    tiles_y = int(np.ceil(h / tile_size))
+    tiles_x = int(np.ceil(w / tile_size))
+
+    # Create a separate Image visual for each tile
+    tile_list = []
+    for ty in range(tiles_y):
+        for tx in range(tiles_x):
+            # Calculate tile position and size
+            x = tx * tile_size
+            y = ty * tile_size
+
+            w_tile = min(tile_size, w - x)
+            h_tile = min(tile_size, h - y)
+
+            # Skip if tile is empty
+            if w_tile <= 0 or h_tile <= 0:
+                continue
+
+            # Extract tile data
+            tile_data = image[y : y + h_tile, x : x + w_tile]
+
+            tile_list.append(((x, y), tile_data))
```

Comment:
This one's a bit cheeky plus I haven't tested it 😅 so take it or leave it. (Also, will need an itertools import at the top.) But I felt like trying and it came out pretty neat so here it is.

```suggestion
    tile_list = [
        # note: vispy space is transposed (y, x -> x, y) compared to NumPy
        # indexing space; we want the vispy offsets so we swap them here.
        ((x, y), image[y : y + tile_size, x : x + tile_size])
        for y, x in itertools.product(
            range(0, h, tile_size), range(0, w, tile_size)
        )
    ]
```

note btw that we don't need to compute the tile size, because NumPy does the trimming for us:

```python
In [18]: arr = np.random.random(5)

In [19]: arr[3:6]
Out[19]: array([0.25681064, 0.65399036])
```

Reply:
That's a very nice way of avoiding the nested for loop! I'm integrating this and checking that it works as advertised :sweat_smile: